### ISSUE TITLE
SenseVoice: performance optimizations + E2E parity test gap

### ISSUE THREAD

**Issue by SRjoeee** [findit-studio/mlxrs#378]
## Overview

Deep audit of the SenseVoice-Small Rust port (8545 lines, 13 files) against the upstream Python reference (`mlx-audio/sensevoice.py`, 598 lines). Architecture verified by two independent review teams (mimo-v2.5-pro + Claude Haiku) — **0 correctness bugs found**. Three performance optimization opportunities identified.

## Architecture Correctness: ✅ A+

17 verification points, 15 exact MATCH, 2 acceptable divergences (stricter safety checks):

| # | Area | Verdict |
|---|------|---------|
| 1 | SinusoidalPositionEncoder | MATCH — Rust builds `(T,half)` then broadcasts; Python builds `(B,T,half)` directly |
| 2 | PositionwiseFeedForward / ReLU | MATCH — `maximum(x, 0)` = `nn.relu` |
| 3 | QKV split | MATCH — `split_sections([n_feat, 2*n_feat])` = `split(3)` |
| 4 | FSMN memory (pad + Conv1d) | MATCH |
| 5 | EncoderLayerSANM normalize_before | **Acceptable divergence** — hardcodes pre-norm, rejects `false` at load |
| 6 | sqrt(output_size) scale | MATCH |
| 7 | log_softmax | MATCH — `x - logsumexp(x)` |
| 8 | build_query embedding gather | MATCH — concat order identical |
| 9 | generate / utterance_log_probs | MATCH |
| 10 | greedy_collapse argmax dtype | MATCH — mlx argmax returns U32 (verified in `ops/misc.rs:93-94`) |
| 11 | rich_info id maps | MATCH — all lid/emo/event maps identical |
| 12 | sanitize | **Acceptable divergence** — Rust rejects duplicate keys (stricter) |
| 13 | post_load_hook precedence | MATCH — SPM → tokens.json → id_join |
| 14 | Config defaults | MATCH |
| 15 | LAYERNORM_EPS = 1e-5 | MATCH |
| 16 | Waveform pre-scale = 2^15 | MATCH |
| 17 | ReLU semantics | MATCH |

### 1. Precompute Sinusoidal PE — HIGH impact

`SinusoidalPositionEncoder` (encoder.rs:302-393) recomputes `arange → exp → sin → cos → concat → pad → reshape → cast → add` (~10 lazy graph nodes) on **every** forward call. The encoding depends only on `(T, D)` where `D` is fixed per block (560 for encoders0, 512 for the rest).

**Fix**: Cache a `(1, max_T, D)` encoding at construction, slice to `(1, T, D)` per forward. Store two cached encodings (one per block width).

**Reference**: Whisper precomputes `sinusoids(n_ctx, n_state)` once at construction (encoder.rs:77, 118, 274).

### 2. Batch frame_argmax — MEDIUM-HIGH impact

`rich_info` (model.rs:502-524) calls `frame_argmax` 3 times, each doing:

```
slice → reshape → argmax → item::<u32>()  // GPU sync per call
```

That's **3 separate GPU→CPU syncs** for 3 argmax operations over adjacent rows.

**Fix**: Slice frames 0..3 → `(3, vocab)`, single `argmax(axis=1)` → `(3,)`, one `as_slice::<u32>()`:

```rust
let query = slice(log_probs, [0,0], [3,vocab])?;
let mut preds = argmax(&query, Some(1), false)?;
eval(&[&preds])?;  // single GPU sync
let ids = preds.as_slice::<u32>()?;
// lid_label(ids[0]), emotion_label(ids[1]), event_label(ids[2])
```

### 3. Vectorize apply_lfr — MEDIUM impact

`apply_lfr` (frontend.rs:150-265) loops `for i in 0..t_lfr` with slice+reshape+stack per frame. For a 10-second utterance (~23 LFR frames), this creates ~46 lazy graph ops + one stack.

**Fix**: Replace with a single `ops::indexing::take` using precomputed indices:

```rust
let indices: Vec<i32> = (0..t_lfr)
    .flat_map(|i| (0..lfr_m).map(move |j| (i * lfr_n + j).min(t_padded - 1)))
    .collect();
let idx = Array::from_slice(&indices, &[t_lfr, lfr_m])?;
let gathered = ops::indexing::take(&padded, &idx, 0)?;  // (t_lfr, lfr_m, D)
reshape(&gathered, &[t_lfr, lfr_width])?  // (t_lfr, lfr_m*D)
```

### Already optimized (no action needed)

- **greedy_collapse**: Follows Whisper #362 pattern (single `as_slice` GPU sync, host-side loop)
- **ctc_lo + log_softmax**: Lazy graph, MLX handles fusion
- **Encoder sequential loop**: Non-autoregressive, lazy graph evaluates in one shot
- **CMVN / FSMN / build_query**: Negligible gain from manual fusion

## Missing: E2E Numeric-Parity Test

`mod.rs:45-47` explicitly states: "the gated e2e numeric-parity test against the real `mlx-community/SenseVoiceSmall` checkpoint is a separate change."

This means there is **no verification** that the Rust port produces identical transcription output to the Python reference on real audio. All existing tests use synthetic checkpoints with trivial weights.

**Recommend**: Add a gated E2E test that:
1. Loads the real `mlx-community/SenseVoiceSmall` checkpoint
2. Runs transcription on a known audio sample
3. Compares text + rich-info tags against Python reference output
4. Asserts exact match (or within tolerance for floating-point)

## Test Coverage Summary

| Category | Status |
|----------|--------|
| Shape / closed-form tests | ✅ Every block has oracle tests |
| Synthetic checkpoint tests | ✅ Dense + quantized loader tests |
| E2E numeric-parity (real checkpoint) | ❌ **Missing** |
| Empty / very short audio | ⚠️ Not tested (depends on fbank) |
| Long audio i32 overflow | ✅ `Config::validate` has `checked_mul` + `MAX_CONFIG_CARDINALITY` |
| normalize_before=false | ✅ Rejected at load |
| Duplicate sanitize keys | ✅ Returns `InvariantViolation` |

## Files

- `mlxrs/src/audio/stt/models/sensevoice/mod.rs` (66 lines)
- `mlxrs/src/audio/stt/models/sensevoice/config.rs` (803 lines)
- `mlxrs/src/audio/stt/models/sensevoice/encoder.rs` (1003 lines)
- `mlxrs/src/audio/stt/models/sensevoice/frontend.rs` (372 lines)
- `mlxrs/src/audio/stt/models/sensevoice/model.rs` (889 lines)
- `mlxrs/src/audio/stt/models/sensevoice/loader.rs` (576 lines)
- `mlxrs/src/audio/stt/models/sensevoice/tokenizer.rs` (364 lines)
- Plus 6 test files (4472 lines total)

## Review Methodology

- **Team 1** (mimo-v2.5-pro): Architecture correctness — 17 verification points, line-by-line comparison vs Python reference
- **Team 2** (mimo-v2.5-pro): Performance optimization — analyzed 12 areas against Whisper optimization patterns
- **Team 3** (Claude Haiku): Independent adversarial validation — confirmed all findings, attempted to find missed bugs (none found)
- **Cross-validation**: Two independent models reached identical conclusions

**Comment by SRjoeee** [comment]
2026-06-10 full-model audit follow-up (new findings not in this issue's scope):

- **dtype contract incomplete (P2)**: LFR features are always F32 and the f32 relu zero scalar pierces the existing cast guards — on an fp16/bf16/quantized checkpoint the whole encoder silently promotes to F32 (same class as whisper `1723a5c`, latent today because the official checkpoint is fp32).
- **Shared `greedy_ctc_transcribe` 64Mi logits budget (P2)**: sized for vocab~32; at SenseVoice's vocab=25055 it caps the direct path at ~2.7 minutes of audio.

Full evidence: `MODEL_AUDIT_2026-06-10.md` (branch `fix/whisper-mel-dtype`) §sensevoice. Tracking: #421

**Comment by SRjoeee** [comment]
## 2026-07-08 benchmark update: one-shot long-audio path scales poorly

Follow-up from the full benchmark pass. This is **not** a correctness claim about the transcript content; it is a runtime/long-audio API observation. The long input used here appears to be a poor ASR segment (Silero VAD reports 0 segments and SenseVoice text preview is empty), so the empty text should not be treated as a port correctness bug. The actionable finding is that the current one-shot SenseVoice path lets callers feed long audio directly into the full encoder, and cost degrades sharply with duration.

### Evidence

Model:

```text
/Volumes/T7 Shield/models_local/SenseVoiceSmall-MLX
```

Original benchmark input:

```text
/Volumes/T7 Shield/mlxrs-bench-inputs/hh0329_90_mix_mono16k.wav
1072.0s, 16 kHz mono
```

Original 1072s run:

```text
Full transcription: 63172.6 ms
Feature extraction:   502.8 ms
Encoder forward:    66501.4 ms
Speed: 17.0x realtime
Transcription preview: empty
```

To check whether this was an isolated full-file artifact, I cut prefixes from the same long wav and reran the same benchmark harness with `--runs 1 --warmup 0`:

```text
43.5s excerpt:
  full transcription = 1667.2 ms
  feature extraction =   30.1 ms
  encoder forward    =  435.0 ms
  text preview       = empty

120s excerpt:
  full transcription = 1674.9 ms
  feature extraction =   52.9 ms
  encoder forward    = 1558.2 ms
  text preview       = empty

300s excerpt:
  full transcription = 6376.1 ms
  feature extraction =  135.5 ms
  encoder forward    = 6326.0 ms
  text preview       = empty

1072s full:
  full transcription = 63172.6 ms
  feature extraction =   502.8 ms
  encoder forward    = 66501.4 ms
  text preview       = empty
```

### Interpretation

- Feature extraction remains small and roughly linear.
- Encoder forward dominates and worsens with long sequence length.
- This is consistent with feeding the full long sequence into the SANM/encoder path rather than chunking/windowing first.
- Because the input is likely not a good speech segment, I would not use the empty text as evidence of an ASR correctness bug. The performance/API issue stands independently: callers can currently pass long audio into the one-shot path, and the implementation has no chunk/window strategy or duration guard analogous to wav2vec2's explicit 60s cap.

### Suggested follow-up

1. Add a SenseVoice chunk/window benchmark:
   - one-shot 1072s
   - 30s / 60s / 120s chunked
   - compare wall time, peak memory, and transcript stitching behavior
2. Decide whether the public high-level `transcribe` API should:
   - expose an official chunked long-audio helper, or
   - document/guard max recommended duration for one-shot use.
3. Add a local gated regression/perf fixture if we want to track this across future refactors.

Priority: P1 performance/product gap for long-media workloads, not a confirmed transcript correctness bug.
